shared/actions/chat/attachment.js

```diff
@@ -270,89 +329,56 @@ function* onSelectAttachment({payload: {input}}: Constants.SelectAttachment): Ge
-        curKey = keyChanged.payload.newMessageKey
-      } else if (progress) {
-        const {bytesComplete, bytesTotal} = progress.params
-        yield put(Creators.uploadProgress(curKey, bytesComplete / bytesTotal))
-      } else if (finished) {
-        if (finished.error) {
-          yield put(
-            Creators.updateTempMessage(
-              conversationIDKey,
-              {
-                messageState: 'failed',
-                failureDescription: 'upload unsuccessful',
-              },
-              outboxIDKey
-            )
-          )
-        }
-        yield put(Creators.uploadProgress(curKey, null))
-        yield cancel()
-      }
-    }
-  })
+  // TODO This is really hacky, should get reworked
+  let curKey: ?Constants.MessageKey = null
+  let outboxID = null
 
-  const uploadStart = yield Saga.takeFromChannelMap(channelMap, 'chat.1.chatUi.chatAttachmentUploadStart')
-  uploadStart.response.result()
+  const getCurKey = () => curKey
+  const getOutboxIdKey = () => outboxID
 
-  const previewUploadStart = yield Saga.takeFromChannelMap(
-    channelMap,
-    'chat.1.chatUi.chatAttachmentPreviewUploadStart'
+  const setCurKey = nextCurKey => {
+    curKey = nextCurKey
+  }
+  const setOutboxIdKey = nextOutboxID => (outboxID = nextOutboxID)
+
+  const postAttachment = new EngineRpc.EngineRpcCall(
+    postAttachmentSagaMap(conversationIDKey, preview, title, filename, getCurKey, setCurKey, setOutboxIdKey),
+    ChatTypes.localPostFileAttachmentLocalRpcChannelMap,
+    `localPostFileAttachmentLocal-${conversationIDKey}-${title}-${filename}`,
+    {param}
   )
-  previewUploadStart.response.result()
-  const previewUploadDone = yield Saga.takeFromChannelMap(
-    channelMap,
-    'chat.1.chatUi.chatAttachmentPreviewUploadDone'
+
+  const keyChangedTask = yield Saga.safeTakeEvery(
+    action => action.type === 'chat:outboxMessageBecameReal' && action.payload.oldMessageKey === getCurKey(),
+    function*(action) {
+      curKey = action.payload.newMessageKey
+    }
   )
-  previewUploadDone.response.result()
-  const uploadDone = yield Saga.takeFromChannelMap(channelMap, 'chat.1.chatUi.chatAttachmentUploadDone')
-  uploadDone.response.result()
 
-  yield join(progressTask)
-  Saga.closeChannelMap(channelMap)
-  // The message is updated when uploading finishes via an incoming attachmentuploaded message.
+  try {
+    const result = yield call(postAttachment.run)
+    if (EngineRpc.isFinished(result)) {
+      if (result.error) {
+        const outboxIDKey = yield call(getOutboxIdKey)
+        yield put(
+          Creators.updateTempMessage(
+            conversationIDKey,
+            {
+              messageState: 'failed',
+              failureDescription: 'upload unsuccessful',
+            },
+            outboxIDKey
+          )
+        )
+      }
+      const curKey = yield call(getCurKey)
+      yield put(Creators.uploadProgress(curKey, null))
+    } else {
+      console.warn('Upload Attachment Failed')
```

Comment:
In what case could this be hit? Probably should reset upload progress in this branch as well if the upload is no longer occurring. I guess we weren't handling this kind of exception in the previous code.